OAuth: add possibility to extract roles from access_token

GreenRover opened the pull request:
Not all Oauth implementations (for example Keycloak) provide the groups at the user_info. Those are always at the access_token

Closes #22631

GreenRover committed:
OAuth: add possibility to extract roles from access_token

Not all Oauth implementations (for example Keycloak) provide the groups at the user_info. Those are always at the access_token

Closes #22631

oddlittlebird commented on docs/sources/auth/generic-oauth.md:
```diff
@@ -54,6 +54,9 @@ Check for the presence of a role using the [JMESPath](http://jmespath.org/exampl
 See [JMESPath examples](#jmespath-examples) for more information.
 
+The `role_attribute_path` JMESPath expression will be applied first on the id_toke and userinfo by api_url, like all other options like `email_attribute_path`.
+If there is no hit it will look up the `access_token`.
```
What do you mean by "no hit"?

Also, unclear antecedent. What is the "it" that will look up the access token? Please replace with a more specific noun if possible.

GreenRover replied:
Is the new sentence better?

The `role_attribute_path` JMESPath expression is first applied on the id_token and userinfo by api_url, like all other options. For example, `email_attribute_path`.
If there is no hit it will look up the `access_token`.	If there is no hit, then the JMESPath expression will be applied to the `access_token`.

oddlittlebird replied:
Still not sure what you mean by "no hit." Do you mean if the expression does not find a match?

GreenRover replied:
Yes that is what i mean. Sorry english is not my native language.

GreenRover replied:
@oddlittlebird Is this more clear?

The `role_attribute_path` JMESPath expression is first applied on the id_token and userinfo by api_url, like all other options. For example, `email_attribute_path`.
If the `role_attribute_path` JMESPath expression has no match in the id_toke or userinfo, then the expression will be applied to the `access_token`.

oddlittlebird replied:
Yes, much better!

oddlittlebird commented on docs/sources/auth/generic-oauth.md:
```diff
@@ -54,6 +54,9 @@ Check for the presence of a role using the [JMESPath](http://jmespath.org/exampl
 See [JMESPath examples](#jmespath-examples) for more information.
 
+The `role_attribute_path` JMESPath expression will be applied first on the id_toke and userinfo by api_url, like all other options like `email_attribute_path`.
+If there is no hit it will look up the `access_token`.
```
```suggestion
If there is no hit, then the it will look up the `access_token`.
```

oddlittlebird commented on docs/sources/auth/generic-oauth.md:
```diff
@@ -54,6 +54,9 @@ Check for the presence of a role using the [JMESPath](http://jmespath.org/exampl
 
 See [JMESPath examples](#jmespath-examples) for more information.
 
+The `role_attribute_path` JMESPath expression will be applied first on the id_toke and userinfo by api_url, like all other options like `email_attribute_path`.
```
```suggestion
The `role_attribute_path` JMESPath expression is first applied on the id_token and userinfo by api_url, like all other options. For example, `email_attribute_path`.
```

GreenRover committed:
Improve gramma in doku

Co-Authored-By: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

GreenRover committed:
Try to improve doku spelling again

GreenRover commented:
@aknuds1 I saw you self-assigned this to you a couple of days ago, and I've addressed the comments made by @oddlittlebird since then. Is this good to merge now?

aknuds1 commented:
@GreenRover It has to be reviewed first. I'm too busy for the time being, so it will have to wait until someone else steps up.

levkar commented:
@GreenRover 
We are using keycloak and we can send role in userinfo by creating a mapper under client and enable "add to user info".

aknuds1 commented:
@levkar Thanks for your that info!

aknuds1 commented:
Had a look at the code now - it looks OK technically, but I need input from my colleagues on whether we want the functionality.

aknuds1 commented:
@GreenRover Can you use @levkar's workaround?

GreenRover commented:
@levkar  We use red hat sso in an enterprise flavor. That work around is not possible.

GreenRover commented:
@levkar This is in general completly correckt what red hat sso does.
Because this is what the specificatzion says.

User Info is for informations like:
- Realname
- Email
- Telefone numbers
- Address 
- Company organisation structures
- .....

Access Token:
- roles
- aud (audience)
- realm_access (id provider used)

levkar commented:
I am not knowledgeable about redhat sso. I assumed you are using keycloak.

GreenRover commented:
RHSSO is under the hood keycloak, in an old version and some small differences. But in my case i have to deal with an big corporate installation whit allot of restrictions.

GreenRover commented:
@levkar is there a reason that speaks against this function?

aknuds1 commented:
@GreenRover Checking with my colleagues if we should accept this feature or not.

levkar commented:
I am not against this function. I just answered your first message.

> Not all Oauth implementations (for example Keycloak) provide the groups at the user_info. Those are always at the access_token

ch-egli commented:
We use Keycloak as well and this feature would be highly appreciated. Is there any reason to not approve it?

mr-yaky commented:
I had the same problem for `roles` with Keycloak and then I have used the recommendation from @levkar 
>We are using keycloak and we can send role in userinfo by creating a mapper under client and enable "add to user info".

After doing this workaround I finally solved the problem. Thanks!

GreenRover commented:
@aknuds1 Are there any updates?

marefr commented:
@GreenRover Thank you for your contribution. However, after consideration we've decided to close this for now. 

Given that you're the only one currently using a fork of Keycloak we cannot accept it. Yes, there are others using Keycloak, but you can configure this in Keycloak. I understand it's probably hard to migrate away from that old version, but to accept this we would need some more proof that it's a feature needed by many users.

We'd like to suggest that you and others keep the discussion going in the feature request you opened (#22631), and ask for the community to vote on it, explaining their use case and specifying their specific OAuth provider and possible version. Just because we're closing it doesn't mean we cant reopen it later.

GreenRover commented:
@marefr Its not an old version of Keycloak. It is:
https://access.redhat.com/products/red-hat-single-sign-on
This is based on an old version of keycloak. In generel RH SSO is some month-years behind the open source community..

marefr commented:
@GreenRover yeah I understand that. Did describe that badly. But our view of this is the same. You're the only one so far using Red Hat Single Sign-On and having this problem.

ch-egli commented:
I just don't get your argumentation: 
Of course it is possible to add roles to the ID token or /userinfo endpoint, but if you have no influence on the OAuth infrastructure of your enterprise you are lost. 
And the OpenID specification does not provide a "roles" claim in the ID token (see https://openid.net/specs/openid-connect-core-1_0.html#IDToken and https://openid.net/specs/openid-connect-core-1_0.html#Claims), so the Grafana mechanism for authorization relies on optional values that are not necessarily present in an OAuth infrastructure. 
I am not sure where to address this, but I would expect Grafana to either accept a pull-request like this one or provide an other viable solution, thank you.

marefr commented:
@ch-egli thanks. The point here is that to "accept a pull-request like this one or provide an other viable solution" we need indicators this is a common use case and needed by many users (more than 2). Can I suggest you to continue the discussions in #22631 and report what's your OAuth provider and possible version.

ch-egli commented:
@marefr Thanks for your reply. My point is based on the OIDC specification and not on a specific OAuth provider. But I will address it in #22631...
